src/raft/read_only.rs

```diff
@@ -0,0 +1,143 @@
+    fn default() -> ReadOnlyOption {
+        ReadOnlyOption::Safe
+    }
+}
+
+// ReadState provides state for read only query.
+// It's caller's responsibility to send MsgReadIndex first before getting
+// this state from ready. It's also caller's duty to differentiate if this
+// state is what it requests through request_ctx, e.g. given a unique id as
+// request_ctx.
+#[derive(Default, Debug, PartialEq, Clone)]
+pub struct ReadState {
+    pub index: u64,
+    pub request_ctx: Vec<u8>,
+}
+
+#[derive(Default, Debug, Clone)]
+pub struct ReadIndexStatus {
+    pub req: Message,
+    pub index: u64,
+    pub acks: HashSet<u64>,
+}
+
+#[derive(Default, Debug, Clone)]
+pub struct ReadOnly {
+    pub option: ReadOnlyOption,
+    pub pending_read_index: HashMap<Vec<u8>, ReadIndexStatus>,
+    pub read_index_queue: VecDeque<Vec<u8>>,
+}
+
+impl ReadOnly {
+    pub fn new(option: ReadOnlyOption) -> ReadOnly {
+        ReadOnly {
+            option: option,
+            pending_read_index: HashMap::new(),
+            read_index_queue: VecDeque::new(),
+        }
+    }
+
+    /// add_request adds a read only request into readonly struct.
+    /// `index` is the commit index of the raft state machine when it received
+    /// the read only request.
+    /// `m` is the original read only request message from the local or remote node.
+    pub fn add_request(&mut self, index: u64, m: Message) {
+        let ctx = m.get_entries()[0].get_data().to_vec();
+        if self.pending_read_index.contains_key(&ctx) {
+            return;
+        }
+        let status = ReadIndexStatus {
+            req: m,
+            index: index,
+            acks: HashSet::new(),
+        };
+        self.pending_read_index.insert(ctx.clone(), status);
+        self.read_index_queue.push_back(ctx);
+    }
+
+    /// rev_ack notifies the ReadOnly struct that the raft state machine received
+    /// an acknowledgment of the heartbeat that attached with the read only request
+    /// context.
+    pub fn recv_ack(&mut self, m: &Message) -> usize {
+        match self.pending_read_index.get_mut(m.get_context()) {
+            None => 0,
+            Some(rs) => {
+                rs.acks.insert(m.get_from());
+                // add one to include an ack from local node
+                rs.acks.len() + 1
+            }
+        }
+    }
+
+    /// advance advances the read only request queue kept by the ReadOnly struct.
+    /// It dequeues the requests until it finds the read only request that has
+    /// the same context as the given `m`.
+    pub fn advance(&mut self, m: &Message) -> Vec<ReadIndexStatus> {
+        let mut rss = vec![];
+        if let Some(i) = self.read_index_queue.iter().position(|x| {
+            if !self.pending_read_index.contains_key(x) {
+                panic!("cannot find correspond read state from pending map");
+            }
+            *x == m.get_context()
+        }) {
+            let to_keep = self.read_index_queue.split_off(i + 1);
```

Comment:
I don't get the point. It will still allocates a new buffer rewriting a `split_off` using `pop_front`. Without a new vector, where does it put the two collections after split.